src/storage-client/src/controller.rs

```diff
@@ -2010,14 +2019,70 @@ where
             data_shard,
         } = new_metadata;
 
-        // Update in memory collection metadata.
-        let metadata = self
-            .state
-            .collections
-            .get_mut(&id)
-            .expect("id {:?} must exist");
-        metadata.collection_metadata.data_shard = data_shard;
-        metadata.collection_metadata.remap_shard = remap_shard;
+        // Update in memory collection metadata if the collection has been
+        // registered.
+        if let Some(metadata) = self.state.collections.get_mut(&id) {
+            metadata.collection_metadata.data_shard = data_shard;
+            metadata.collection_metadata.remap_shard = remap_shard;
+
+            if replace_data_shard {
+                let persist_client = self
+                    .persist
+                    .lock()
+                    .await
+                    .open(self.persist_location.clone())
+                    .await
+                    .unwrap();
+
+                let purpose = format!("controller data {}", id);
+                let write = persist_client
+                    .open_writer(
+                        data_shard,
+                        &purpose,
+                        Arc::new(metadata.collection_metadata.relation_desc.clone()),
+                        Arc::new(UnitSchema),
+                    )
+                    .await
+                    .expect("invalid persist usage");
+
+                // Construct the handle in a separate block to ensure all error paths are diverging
+                let since_handle = {
+                    let mut handle: SinceHandle<_, _, _, _, PersistEpoch> = persist_client
+                        .open_critical_since(
+                            data_shard,
+                            PersistClient::CONTROLLER_CRITICAL_SINCE,
+                            &purpose,
+                        )
+                        .await
+                        .expect("invalid persist usage");
+
+                    let since = handle.since().clone();
+
+                    // When rewriting data, must be guaranteed to be able to
```

Comment:
Why is this different from the other code that fences and registers handles? We already correctly rewrote the metadata (which should also only succeed when we're the leader), but someone else might have become the leader in between that happening and us updating the in-memory handles. But it seems in that case we can also just `halt!` out like the other code does?

Reply:
I whipped up a commit that shows what I mean: https://github.com/aljoscha/materialize/commit/12be209d88e706aa14fdc193110488c30c41c704. If we like that, feel free to use that one.

Reply:
Ah, I had misread the `loop` in that statement; I realize now it's just ensuring that the epoch hasn't changed since we opened the critical since handle. Not sure what my misread was before. At any rate, thank you for the commit.